rho/ansible_utils.py

```diff
@@ -0,0 +1,165 @@
+
+    Returns after the process completes.
+
+    :param cmd_string: the command to run.
+    :param vault_pass: the password to the user's Ansible Vault.
+    :param env: the environment to run the subprocess in.
+    :param log_path: a path to write the process's log to. Defaults to
+        'XDG_DATA_HOME/rho/ansible_log'.
+    :param log_to_stdout: if True, write Ansible's log to stdout. Defaults to
+        True.
+    :param ansible_verbosity: the number of v's of Ansible verbosity.
+    :returns: the popen.spawn object for the process.
+    """
+
+    # pexpect provides the ability to send the process's output to a
+    # single Python file object. We want to send it to a file and
+    # maybe also stdout. The solution is to have pexpect log to the
+    # file and then use 'tail -f' to copy that to stdout.
+
+    if not log_path:
+        log_path = ANSIBLE_LOG_PATH
+
+    if ansible_verbosity:
+        cmd_string = cmd_string + ' -' + 'v' * ansible_verbosity
+
+    result = None
+    try:
+        utilities.ensure_data_dir_exists()
+        with open(log_path, 'wb') as logfile:
+            pass
+        with open(log_path, 'r+b') as logfile:
+            log.debug('Running Ansible: %s', cmd_string)
+            child = pexpect.spawn(cmd_string, timeout=None,
+                                  env=env)
+
+            if log_to_stdout:
+                utilities.threaded_tailing(log_path, ansible_verbosity)
+
+            result = child.expect('Vault password:')
+            child.sendline(vault_pass)
+            # Set the log file *after* we send the user's Vault
+            # password to Ansible, so we don't log the password.
+            child.logfile = logfile
+            last_pos = logfile.tell()
+
+            i = child.expect([pexpect.EOF, 'Enter passphrase for key .*:',
+                              'you want to continue connecting (yes/no)?'])
+            while i:
+                new_pos = logfile.tell()
+                logfile.seek(last_pos)
+                logfile_lines = logfile.readlines()
+                log.info(logfile_lines)
+                print(logfile_lines[-1].replace('\r\n', ''))
+                logfile.seek(new_pos)
+                last_pos = new_pos
+                child.logfile = None
+                # Ansible has already printed a prompt; it would be
+                # confusing if getpass printed another one.
+                child.sendline(getpass(''))
+                child.logfile = logfile
+                i = child.expect([pexpect.EOF, 'Enter passphrase for key .*:',
+                                  'you want to continue connecting (yes/no)?'])
+
+            if child.isalive():
+                child.wait()
+            if log_to_stdout:
+                time.sleep(2)
+
+        return child
+    except pexpect.EOF:
+        print(str(result))
+        print('pexpect unexpected EOF')
+    except pexpect.TIMEOUT:
+        print(str(result))
+        print('pexpect timed out')
```

Comment:
Is this user facing, or just a developer debug output? I don't know if users would get that much out of this message, is there a way to be more explicative about what would cause these errors?

Reply:
Took another look at that section. These exceptions occur if pexpect fails to find a match for one of the expect statements. This would happen if Ansible didn't prompt for a Vault password or didn't prompt for an SSH key password when we expected one.

Ideally we would handle those cases better, but that's a bigger change that I would like to keep out of this refactoring PR. For this change, I'll update the error messages to be more informative.